### PR TITLE
[SYCLomatic] Support fuzzy match of --query-api-mapping.

### PR DESCRIPTION
**Pull request by tangjj11**
Signed-off-by: Tang, Jiajun jiajun.tang@intel.com

### CODE REVIEW

**Review by yihwang-nv — APPROVED** [review]

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.h [hunk 1]
```diff
@@ -25,13 +27,14 @@ class APIMapping {
   static void initEntryMap();
 
   template <class StreamTy>
-  static void queryAPIMapping(StreamTy &Stream, const std::string &Key) {
+  static void queryAPIMapping(StreamTy &Stream, std::string &Key) {
+    Key.erase(0, Key.find_first_not_of(" "));
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APIMapping/QueryAPIMapping.h]: use std::trim to refine the code.

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.h [hunk 2]
```diff
@@ -25,13 +27,14 @@ class APIMapping {
...
+    Key.erase(Key.find_last_not_of(" ") + 1);
     auto Iter = EntryMap.find(Key);
-    if (Iter == EntryMap.end() || Iter->second == "NA")
+    if (Iter == EntryMap.end())
       Stream << "The API Mapping is not available\n";
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APIMapping/QueryAPIMapping.h]: how about search again with key tolower(key)

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.h (hunk 1 above)
Comment (tangjj11) [reply clang/lib/DPCT/APIMapping/QueryAPIMapping.h]: C++ not have this function.

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.h (hunk 2 above)
Comment (tangjj11) [reply clang/lib/DPCT/APIMapping/QueryAPIMapping.h]: fixed

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp [hunk 3]
```diff
@@ -8,26 +8,55 @@
...
+std::unordered_map<std::string, size_t> APIMapping::EntryMap;
+std::vector<llvm::StringRef> APIMapping::EntryArray;
 
-void APIMapping::registerEntry(llvm::StringRef Name,
-                               llvm::StringRef SourceCode) {
-  EntryMap[Name] = SourceCode;
+void APIMapping::registerEntry(std::string Name, llvm::StringRef SourceCode) {
+  std::replace(Name.begin(), Name.end(), '$', ':');
+  // Try to fuzz the original API name:
+  // 1. Remove 0/1/2/... leader '_'.
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp]:
```suggestion
  // 1. Remove partial or all leading '_'. 
```

Comment (zhimingwang36) [reply clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp]: for alias, please also consider add fuzzing case:  API_NAME_V1, API_NAME_V2

Comment (tangjj11) [reply clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp]: fixed, and no API name has `_v1`, so only add `_v2` fuzz.

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**